syntax = "proto3";
import "functions.proto";

message HSV {
    float h = 1;
    float s = 2;
    float v = 3;
}

message ConstColorEffectConfig {
    HSV color = 1;
}

message RainbowEffectConfig {
    FloatFunction hue_start = 1;
    FloatFunction hue_end = 2;
}

message EffectConfig {
    uint32 start_time = 1;
    uint32 end_time = 2;
    string segments = 3;

    float repeat_num = 4;
    float repeat_start = 5;
    float repeat_end = 6;
}

message EffectProto {

    EffectConfig effect_config = 1;

    oneof effect {
        ConstColorEffectConfig const_color = 2;
        RainbowEffectConfig rainbow = 3;
    }
}

message AnimationProto {
    repeated EffectProto effects = 1;
    uint32 duration_ms = 2;
    uint32 num_repeats = 3; // 0 means forever
}
